### PR TITLE
Use service bind for searching LDAP groups

### PR DESCRIPTION
**Pull request by mitchelldavis**
Addressing #2387!

### CODE REVIEW

**Comment by jefferai** [comment]
@shomron Any chance I can get your opinion on this?

File: builtin/credential/ldap/backend.go [hunk 1]
```diff
@@ -122,6 +122,17 @@ func (b *backend) Login(req *logical.Request, username string, password string)
 		return nil, logical.ErrorResponse(fmt.Sprintf("LDAP bind failed: %v", err)), nil
 	}
 
+	// We re-bind to the BindDN if it's defined because we assume
+	// the BindDN should be the one to search, not the user logging in.
+	if cfg.DiscoverDN || (cfg.BindDN != "" && cfg.BindPassword != "") {
+		if err := c.Bind(cfg.BindDN, cfg.BindPassword); err != nil {
+			return nil, logical.ErrorResponse(err.Error()), nil
```
Comment (shomron) [inline builtin/credential/ldap/backend.go]: Given we are performing up to three binds per login - I think we should improve the error messages to more clearly indicate where in the process we failed.

File: builtin/credential/ldap/backend.go [hunk 2]
```diff
@@ -122,6 +122,17 @@ func (b *backend) Login(req *logical.Request, username string, password string)
...
+		}
+		if b.Logger().IsDebug() {
+			b.Logger().Debug("auth/ldap: Re-Bound to original BindDN")
+		}
+	}
+
 	userDN, err := b.getUserDN(cfg, c, bindDN)
```
Comment (shomron) [inline builtin/credential/ldap/backend.go]: Between `bindDN` and `cfg.BindDN`, following the flow is a bit challenging. Perhaps we should rename the `bindDN` local variable to something like `userBindDN`, to differentiate it from the _service_ bind DN - `cfg.BindDN` ?

Also note `userBindDN` would not necessarily equal `userDN` - userDN is the canonical representation of the user, whereas `userBindDN` could in fact be using a different mechanism such as user principal name (`user@domain`).

**Comment by jackdwyer** [comment]
@mitchelldavis thanks for addressing this, I just started working on JumpCloud LDAP integration with our Vault cluster :+1:

File: builtin/credential/ldap/backend.go [hunk 3]
```diff
@@ -122,6 +122,17 @@ func (b *backend) Login(req *logical.Request, username string, password string)
 		return nil, logical.ErrorResponse(fmt.Sprintf("LDAP bind failed: %v", err)), nil
 	}
 
+	// We re-bind to the BindDN if it's defined because we assume
+	// the BindDN should be the one to search, not the user logging in.
+	if cfg.DiscoverDN || (cfg.BindDN != "" && cfg.BindPassword != "") {
```
Comment (mitchelldavis) [inline builtin/credential/ldap/backend.go]: @shomron, should I just get rid of `cfg.DiscoverDN` clause?  And simply re-bind if the second clause is true?  That makes the most sense to me.

File: builtin/credential/ldap/backend.go (hunk 1 above)
Comment (mitchelldavis) [inline builtin/credential/ldap/backend.go]: So maybe:
```
return nil, logical.ErrorResponse("Encountered an error when re-binding as the BindDN User: " + err.Error()), nil
```

File: builtin/credential/ldap/backend.go (hunk 2 above)
Comment (mitchelldavis) [inline builtin/credential/ldap/backend.go]: That would make sense to me, but if we're going to go down that path, shouldn't we rename the `getBindDN` method then?

**Comment by mitchelldavis** [comment]
@shomron, I believe I've addressed all that.  Does the code reflect what you were thinking?

**Comment by lattwood** [comment]
@shomron @mitchelldavis anything I can do to help get this into the next release?

**Comment by mitchelldavis** [comment]
I'm not sure @lattwood.  I think we're just waiting for a final review.

**Comment by jefferai** [comment]
@shomron thanks for reviewing this!

@shomron @mitchelldavis any backwards incompatible or notably changed behavior as a result of this?

In addition, any necessary documentation changes?

**Comment by mitchelldavis** [comment]
@jefferai, @shomron, that build issue looks like the shamir unit tests failed.  Have we seen that happening in other places because I didn't touch that code.

@jefferai, to answer your question about the documentation - The documentation didn't really specify much to begin with.  It makes sense to me to mention that the binddn user needs permission to search groups because it's what is being used to do that search.

**Comment by jefferai** [comment]
@mitchelldavis We get some transient unit test failures that we know of (often issues with the MySQL docker container) so don't worry about it if it's not a part of your changes.

Please add the mention in the docs about the binddn user needing permission to search groups.

Still waiting to hear from you and @shomron about any backwards incompatibilities.

**Comment by mitchelldavis** [comment]
@jefferai, I've added those updates to the docs.  I would love @shomron to look at the few lines I edited/added though.  I'm not an ldap expert so want to make sure that I'm not adding confusion or am outright wrong.

**Comment by lattwood** [comment]
@mitchelldavis might be worth mentioning the user and group searches are done under separate (but identical) binds.

**Comment by mitchelldavis** [comment]
Because we have to bind to the `binddn` user twice @lattwood?

**Comment by shomron** [comment]
The clarification in the documentation looks good to me.

@jefferai I don't see this as a backwards-incompatible change - it might be worthwhile adding something along these lines to the release notes:

_LDAP Auth Backend - Group membership queries will now run as the `binddn` user when `binddn`/`bindpass` are configured, rather than as the authenticating user as was the case previously._

**Comment by jefferai** [comment]
Sounds good to me!

Merging -- thanks everyone!

**Pull request merged**